lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -758,57 +758,62 @@ bool HandshakeBuilder::visitHandshake(handshake::BranchOp op) {
 /// Two outputs conditional branch operation.
 /// Please refer to test_conditional_branch.mlir test case.
 bool HandshakeBuilder::visitHandshake(ConditionalBranchOp op) {
-  ValueVector controlSubfields = portList[0];
+  ValueVector conditionSubfields = portList[0];
   ValueVector argSubfields = portList[1];
-  ValueVector result0Subfields = portList[2];
-  ValueVector result1Subfields = portList[3];
+  ValueVector trueResultSubfields = portList[2];
+  ValueVector falseResultSubfields = portList[3];
 
-  Value controlValid = controlSubfields[0];
-  Value controlReady = controlSubfields[1];
-  Value controlData = controlSubfields[2];
+  Value conditionValid = conditionSubfields[0];
+  Value conditionReady = conditionSubfields[1];
+  Value conditionData = conditionSubfields[2];
   Value argValid = argSubfields[0];
   Value argReady = argSubfields[1];
-  Value result0Valid = result0Subfields[0];
-  Value result0Ready = result0Subfields[1];
-  Value result1Valid = result1Subfields[0];
-  Value result1Ready = result1Subfields[1];
-
-  // ConditionalBranch will work only when the control input is active.
-  auto validWhenOp = rewriter.create<WhenOp>(insertLoc, controlValid,
-                                             /*withElseRegion=*/false);
-  rewriter.setInsertionPointToStart(&validWhenOp.thenRegion().front());
-  auto branchWhenOp = rewriter.create<WhenOp>(insertLoc, controlData,
-                                              /*withElseRegion=*/true);
-
-  // When control signal is true, the first branch is selected
-  rewriter.setInsertionPointToStart(&branchWhenOp.thenRegion().front());
-  rewriter.create<ConnectOp>(insertLoc, result0Valid, argValid);
-  rewriter.create<ConnectOp>(insertLoc, argReady, result0Ready);
+  Value trueResultValid = trueResultSubfields[0];
+  Value trueResultReady = trueResultSubfields[1];
+  Value falseResultValid = falseResultSubfields[0];
+  Value falseResultReady = falseResultSubfields[1];
 
-  if (!op.isControl()) {
-    Value argData = argSubfields[2];
-    Value result0Data = result0Subfields[2];
-    rewriter.create<ConnectOp>(insertLoc, result0Data, argData);
-  }
+  auto conditionArgValid = rewriter.create<AndPrimOp>(
+      insertLoc, conditionValid.getType(), conditionValid, argValid);
+
+  auto conditionNot = rewriter.create<NotPrimOp>(
```

Comment:
Nice, I didn't realize `ConditionalBranchOp` only ever has two outputs. That simplifies things a bit. I guess if we ever need to generalize it to variadic outputs, we can use decoders or something along those lines.

Reply:
Exactly!